[12.0] mail_tracking: avoid common users accessing to general trackings

chienandalu opened the pull request:
A user accessing to all trackings with no restrictions could view some sensible informations contained in the mail tracking mail subject. For example, employees leave causes.

Refine which mail tracking record is allowed to see. We can rely on the ACLs of the related records so we minimize inconvenient message subject leaks.

A regular user can read these mail.tracking.email records:

- Those with a linked mail.message that the user can read.
- Those with a linked mail.mail that the user can read.
- Those with no message/mail link but a linked partner that the user can read.
- Those with no linked records.

TT31399

cc @Tecnativa

pedrobaeza commented:
@chienandalu what happened with this?

chienandalu commented:
I have to finish it, yes

pedrobaeza commented on mail_tracking/static/src/js/mail_tracking.js:
```diff
@@ -140,24 +141,28 @@ odoo.define('mail_tracking.partner_tracking', function (require) {
-            };
-            this.do_action(action);
+            // Only admins can check advanced tracking stuff
+            session.user_has_group('base.group_system').then(has_group => {
```
Why this limitation? It's very useful for people to check the tracking lines.

chienandalu replied:
I took a new approach dinamically evaluating acls the same way mail.message does relaying on the related records permissions to filter the trackings a user can see.

chienandalu committed:
[FIX] mail_tracking: permissions

Refine which mail tracking record is allowed to see. We can rely on the
ACLs of the related records so we minimize inconvenient message subject
leaks.

A regular user can read these mail.tracking.email records:

- Those with a linked mail.message that the user can read
- Those with a linked mail.mail that the user can read
- Those with no message/mail link but a linked partner that the user can
read.
- Those with no linked records.

TT31399

pedrobaeza commented:
Please check CI

chienandalu commented:
Fixed

pedrobaeza commented on mail_tracking/models/res_partner.py:
```diff
@@ -26,7 +26,7 @@ def _compute_tracking_emails_count(self):
             count = 0
             if partner.email:
-                count = self.env['mail.tracking.email'].search_count([
+                count = len(self.env['mail.tracking.email']._search([
```
Isn't this going to make very slow the partner view loading? What about simply not putting the number on the smart-button?

chienandalu replied:
Ok, I took the middle path: admins can see the number (they skip ACLs) but regular internal users don't

pedrobaeza commented:
Can you display any banner saying that the shown records may not be all depending on the access rules?

chienandalu commented:
> Can you display any banner saying that the shown records may not be all depending on the access rules?

Can it be done in the tree view or do we have to use a `banner_route`?

pedrobaeza commented:
I can't say, sorry

pedrobaeza commented on mail_tracking/models/mail_tracking_email.py:
```diff
@@ -117,11 +118,83 @@ def write(self, vals):
+        )
+
+    def read(self, fields=None, load="_classic_read"):
+        """Override to explicitely call check_access_rule, that is not called
+        by the ORM. It instead directly fetches ir.rules and apply them."""
```
```suggestion
        """Override to explicitly call check_access_rule, that is not called
        by the ORM. It instead directly fetches ir.rules and apply them.
        """
```

chienandalu replied:
Done

pedrobaeza commented:
/ocabot merge major

pedrobaeza commented:
/ocabot merge major

pedrobaeza commented:
Well, let's fw-port it to v13 and up and let this one as is.

pedrobaeza commented:
/ocabot merge major

pedrobaeza commented:
/ocabot merge major
